joern-cli/frontends/jssrc2cpg/src/main/scala/io/joern/jssrc2cpg/astcreation/AstForStatementsCreator.scala

```diff
@@ -442,4 +437,365 @@ trait AstForStatementsCreator { this: AstCreator =>
+
+    val thisNextNode = createIdentifierNode(iteratorName, forInOfStmt)
+
+    val rhsArgs = List(Ast(thisNextNode))
+    val rhsAst  = callAst(rhsNode, rhsArgs, receiver = Option(nextReceiverNode))
+
+    val doneBaseArgs = List(Ast(lhsNode), rhsAst)
+    val doneBaseAst  = callAst(doneBaseNode, doneBaseArgs)
+    Ast.storeInDiffGraph(doneBaseAst, diffGraph)
+
+    val doneMemberNode = createFieldIdentifierNode("done", forInOfStmt.lineNumber, forInOfStmt.columnNumber)
+
+    val testNode =
+      createFieldAccessCallAst(doneBaseNode, doneMemberNode, forInOfStmt.lineNumber, forInOfStmt.columnNumber)
+
+    val testCallArgs = List(testNode)
+    val testCallAst  = callAst(testCallNode, testCallArgs)
+
+    val whileLoopAst = Ast(whileLoopNode).withChild(testCallAst).withConditionEdge(whileLoopNode, testCallNode)
+
+    // while loop variable assignment:
+    val loopVariableAssignmentAsts = loopVariableNames.map { loopVariableName =>
+      val whileLoopVariableNode = createIdentifierNode(loopVariableName, forInOfStmt)
+      val baseNode              = createIdentifierNode(resultName, forInOfStmt)
+      val memberNode            = createFieldIdentifierNode("value", forInOfStmt.lineNumber, forInOfStmt.columnNumber)
+      val accessAst = createFieldAccessCallAst(baseNode, memberNode, forInOfStmt.lineNumber, forInOfStmt.columnNumber)
+      val variableMemberNode =
+        createFieldIdentifierNode(loopVariableName, forInOfStmt.lineNumber, forInOfStmt.columnNumber)
+      val variableAccessAst =
+        createFieldAccessCallAst(accessAst, variableMemberNode, forInOfStmt.lineNumber, forInOfStmt.columnNumber)
+      val loopVariableAssignmentNode = createCallNode(
+        s"$loopVariableName = $resultName.value.$loopVariableName",
+        Operators.assignment,
+        DispatchTypes.STATIC_DISPATCH,
+        forInOfStmt.lineNumber,
+        forInOfStmt.columnNumber
+      )
+      val loopVariableAssignmentArgs = List(Ast(whileLoopVariableNode), variableAccessAst)
+      callAst(loopVariableAssignmentNode, loopVariableAssignmentArgs)
+    }
+
+    val whileLoopBlockNode = createBlockNode(forInOfStmt)
+    scope.pushNewBlockScope(whileLoopBlockNode)
+    localAstParentStack.push(whileLoopBlockNode)
+
+    // while loop block:
+    val bodyAst = astForNodeWithFunctionReference(forInOfStmt.json("body"))
+
+    val whileLoopBlockChildren = loopVariableAssignmentAsts :+ bodyAst
+    setArgumentIndices(whileLoopBlockChildren)
+    val whileLoopBlockAst = Ast(whileLoopBlockNode).withChildren(whileLoopBlockChildren)
+
+    scope.popScope()
+    localAstParentStack.pop()
+
+    // end surrounding block:
+    scope.popScope()
+    localAstParentStack.pop()
+
+    val blockNodeChildren =
+      Seq(iteratorAssignmentAst, Ast(resultNode)) ++ loopVariableNodes.map(Ast(_)) :+ whileLoopAst.withChild(
+        whileLoopBlockAst
+      )
+    setArgumentIndices(blockNodeChildren)
+    Ast(blockNode).withChildren(blockNodeChildren)
```

Comment:
you could use `blockAst` here

Reply:
Right, good point. The whole thing is target to some refactoring later on anyway. Then I will do this.